daemon/ssh-agent/test-communication.c

```diff
@@ -0,0 +1,453 @@
+	gck_attributes_find_ulong (priv_attrs, CKA_KEY_TYPE, &value);
+	g_assert (value == CKK_RSA);
+	/* public looks reasonable */
+	gck_attributes_find_ulong (pub_attrs, CKA_CLASS, &value);
+	g_assert (value == CKO_PUBLIC_KEY);
+	gck_attributes_find_ulong (pub_attrs, CKA_KEY_TYPE, &value);
+	g_assert (value == CKK_RSA);
+	/* public exponent should be same */
+	priv_attr = gck_attributes_find (priv_attrs, CKA_PUBLIC_EXPONENT);
+	pub_attr = gck_attributes_find (pub_attrs, CKA_PUBLIC_EXPONENT);
+	g_assert (memcmp (priv_attr->value, pub_attr->value, priv_attr->length) == 0);
+
+	/* try to write the public key */
+	ret = gkd_ssh_agent_proto_write_public (&resp, pub_attrs);
+	g_assert (ret);
+	g_assert (egg_buffer_length(&resp) != 0);
+
+	/* Read the written public key */
+	gck_builder_init (&new_pub);
+	offset = 0; /* in this case we do not have message length and operation */
+	ret = gkd_ssh_agent_proto_read_public (&resp, &offset, &new_pub, &algo);
+	g_assert (ret);
+	g_assert (algo == CKK_RSA);
+	new_pub_attrs = gck_builder_end (&new_pub);
+	g_assert (new_pub_attrs);
+
+	/* check that the parameters in old and new public key match */
+	new_pub_attr = gck_attributes_find (new_pub_attrs, CKA_PUBLIC_EXPONENT);
+	pub_attr = gck_attributes_find (pub_attrs, CKA_PUBLIC_EXPONENT);
+	g_assert (memcmp (new_pub_attr->value, pub_attr->value, new_pub_attr->length) == 0);
+
+	new_pub_attr = gck_attributes_find (new_pub_attrs, CKA_MODULUS);
+	pub_attr = gck_attributes_find (pub_attrs, CKA_MODULUS);
+	g_assert (memcmp (new_pub_attr->value, pub_attr->value, new_pub_attr->length) == 0);
+
+	/* cleanup */
+	free (stype);
+	gck_attributes_unref (priv_attrs);
+	gck_attributes_unref (pub_attrs);
+	gck_attributes_unref (new_pub_attrs);
+	egg_buffer_uninit (&buffer);
+	egg_buffer_uninit (&resp);
+}
+
+static void
+test_read_write_dsa (Test *test, gconstpointer unused)
+{
+	GckAttributes *priv_attrs, *pub_attrs, *new_pub_attrs;
+	const GckAttribute *priv_attr, *pub_attr, *new_pub_attr;
+	GckBuilder priv, pub, new_pub;
+	EggBuffer buffer, resp;
+	gsize offset = 5; /* skipping message number (4B length, 1B message) */
+	gchar *stype;
+	gboolean ret;
+	gulong value, algo;
+
+	/* Prepare intercepted message from  ssh-add  */
+	egg_buffer_init_static (&buffer, test->dsakey, test->dsakey_len);
+	egg_buffer_init (&resp, 128);
+
+	/* check the key type */
+	ret = egg_buffer_get_string (&buffer, 5, &offset, &stype, (EggBufferAllocator)g_realloc);
+	g_assert (ret);
+	g_assert (strcmp(stype, "ssh-dss") == 0);
+
+	/* parse the key to PKCS11 structures */
+	gck_builder_init (&pub);
+	gck_builder_init (&priv);
+	ret = gkd_ssh_agent_proto_read_pair_dsa (&buffer, &offset, &priv, &pub);
+	g_assert (ret);
+
+	/* Dump data */
+	pub_attrs = gck_builder_end (&pub);
+	g_assert (pub_attrs);
+	priv_attrs = gck_builder_end (&priv);
+	g_assert (priv_attrs);
+	printf ("Private: %s\n", gck_attributes_to_string (priv_attrs));
+	printf ("Public: %s\n", gck_attributes_to_string (pub_attrs));
```

Comment:
not sure if printing the values is a good/typical approach, but wouldn't printing of private part be enough to reproduce any kind of failures?